documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+
+### Inner parts of a test
+
+While writing test cases, it will be easier to think of the test case's contents as if there were three clear divisions:
+
+- The first lines of the test should deal with setting up the necessary variables and values that will be used for the main functionality that will be tested.
+- The body of the test should focus on using these values to trigger the main functionality that is being tested.
+- The last set of lines should be focused on comparing the results obtained, to confirm that they are in the expected shape, or contain the expected values.
+
+### Using conditionals
+
+Ideally, each test case should only have one possible behavior. To minimize unexpected outcomes, conditionals should be avoided. The only case in which conditionals are encouraged is to specify when to skip a test. Some tests might only run in a browser, or only run in NodeJS, or only run live (no record or playback), for this purpose, conditionals can be used. For example:
+
+```ts
+import { isNode } from "@azure/core-http";
+
+describe("testing the client's basic methods", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  it("should test A #browser", function() {
+    if (isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test B #node", function() {
+    if (!isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test C #live", function() {
+    if (process.env.TEST_MODE) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+});
+```
+
+In case conditionals might appear to be necessary in other scenarios, consider separating the test case into as many test cases as necessary first.
+
+### Using delays
+
+The API methods that we provide should always provide a way for users to know when an operation finishes. For this reason, any method that doesn't immediately respond with a completed operation should use our [Long Running Operations strategy (core-lro)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro).
+
+If any method can't use `core-lro` within reasonable time, and thus we are inclined to use a different delay strategy, we should always wait until the next possible operation is can be fulfilled.
+
+For example, in KeyVault-Keys we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:
+
+```ts
+import { delay } from "@azure/test-utils-recorder";
+// ...
+
+describe("Keys client - restore keys and recover backups", () => {
+  // ...
+  it("can restore a key with a given backup", async function() {
+    // ...
+    await client.createKey(keyName, "RSA");
+    const backup = await client.backupKey(keyName);
+    await client.purgeDeletedKey(keyName);
+    while (true) {
+      try {
+        await client.restoreKeyBackup(backup as Uint8Array);
```

Comment:
Unrelated to this testing sample - why is that the result from .backupKey can't be passed, uncasted, to restoreKeyBackup? Are they not normally paired together?

Reply:
Oh the issue is that backup can be undefined. Would it be better if I use the exclamation sign?

Reply:
Probably my lack of understanding in KeyVault - how does 'backup' ever become 'defined' then (it's const, right?). 

Isn't the point of this to restore a key you've backed up (ie, one that definitely did exist?).

Reply:
it's const, but the API can either return a backup or undefined. Perhaps we should throw if there's no backup! I'll bring this up to the KeyVault team.